File: 1-outbox-and-idempotent-delivery/2-csharp/Controllers/HookController.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Microsoft.AspNetCore.Mvc;
using StackExchange.Redis;

namespace Outbox.Controllers;

/// <summary>receiver-mock: an idempotent webhook receiver. It caches the first
/// successful response with SET ... NX EX and replays it byte-identical, rejects a
/// key reused with a different body (409), and fails the first N attempts per key.</summary>
[ApiController]
public class HookController : ControllerBase
{
    private readonly IConnectionMultiplexer _redis;
    private readonly int _failFirst;
    private readonly TimeSpan _ttl;

    public HookController(IConnectionMultiplexer redis, IConfiguration config)
    {
        _redis = redis;
        _failFirst = config.GetValue("FailFirstNAttempts", 0);
        _ttl = TimeSpan.FromSeconds(config.GetValue("IdemTtlSeconds", 86400));
    }

    [HttpPost("/api/hook")]
    public async Task<IActionResult> Hook()
    {
        var db = _redis.GetDatabase();
        using var reader = new StreamReader(Request.Body);
        var rawBody = await reader.ReadToEndAsync();
        var idemKey = Request.Headers["Idempotency-Key"].ToString();
        if (string.IsNullOrEmpty(idemKey))
            return StatusCode(400, new { statusCode = 400, message = "missing Idempotency-Key" });

        var bodyHash = Sha256(rawBody);
        var cacheKey = $"hook:idem:{idemKey}:{bodyHash}";
        var seenHashKey = $"hook:idem:hash:{idemKey}";

        // Cache hit → replay the first response verbatim; the side-effect does NOT run again.
        var cached = await db.StringGetAsync(cacheKey);
        if (cached.HasValue)
            return Replay(cached!);

        // Same key + different body → 409 (body-hash conflict gate).
        var priorHash = await db.StringGetAsync(seenHashKey);
        if (priorHash.HasValue && priorHash != bodyHash)
            return StatusCode(409, new { statusCode = 409, message = "Idempotency-Key reused with a different body" });

        // Demo retry: count attempts per key; fail the first N.
        var attemptKey = $"hook:attempt:{idemKey}";
        var attempt = await db.StringIncrementAsync(attemptKey);
        await db.KeyExpireAsync(attemptKey, _ttl);
        if (attempt <= _failFirst)
            return StatusCode(503, new { statusCode = 503, message = "simulated failure" });

        // Cache miss → run the side-effect exactly once and capture its result.
        var bodyEcho = TryParse(rawBody);
        var eventId = bodyEcho?["id"]?.GetValue<string>() ?? idemKey;
        var result = new HookResult(true, eventId, idemKey, DateTime.UtcNow.ToString("o"), bodyEcho);
        var payload = JsonSerializer.Serialize(result, Json.Options);

        // SET ... NX EX: only the winner writes; a racing first-time request reads the
        // winner's value instead of double-running the side-effect.
        var won = await db.StringSetAsync(cacheKey, payload, _ttl, When.NotExists);
        await db.StringSetAsync(seenHashKey, bodyHash, _ttl);
        if (!won)
        {
            var winner = await db.StringGetAsync(cacheKey);
            if (winner.HasValue) return Replay(winner!);
        }
        return new ContentResult { StatusCode = 200, Content = payload, ContentType = "application/json" };
    }

    private ContentResult Replay(string body)
    {
        Response.Headers["Idempotent-Replayed"] = "true";
        return new ContentResult { StatusCode = 200, Content = body, ContentType = "application/json" };
    }

    private static JsonNode? TryParse(string raw)
    {
        try { return JsonNode.Parse(raw); }
        catch { return null; }
    }

    private static string Sha256(string s)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(s));
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }
}

File: 1-outbox-and-idempotent-delivery/1-java/src/main/java/com/starci/outbox/receiver/HookController.java
package com.starci.outbox.receiver;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.TimeUnit;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Profile;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RestController;

/**
 * receiver-mock: an idempotent webhook receiver. It caches the first successful
 * response with SET ... NX EX and replays it byte-identical, rejects a key reused
 * with a different body (409), and fails the first N attempts per key to demo retry.
 */
@RestController
@Profile("receiver")
public class HookController {

    private final StringRedisTemplate redis;
    private final ObjectMapper objectMapper;
    private final int failFirst;
    private final long ttlSeconds;

    public HookController(StringRedisTemplate redis, ObjectMapper objectMapper,
                          @Value("${app.fail-first-n-attempts}") int failFirst,
                          @Value("${app.idem-ttl-seconds}") long ttlSeconds) {
        this.redis = redis;
        this.objectMapper = objectMapper;
        this.failFirst = failFirst;
        this.ttlSeconds = ttlSeconds;
    }

    @PostMapping("/api/hook")
    public ResponseEntity<String> hook(@RequestBody(required = false) String rawBody,
                                       @RequestHeader(value = "Idempotency-Key", required = false) String idemKey) throws Exception {
        if (rawBody == null) rawBody = "";
        if (idemKey == null || idemKey.isEmpty()) {
            return json(400, "{\"statusCode\":400,\"message\":\"missing Idempotency-Key\"}", false);
        }

        String bodyHash = sha256(rawBody);
        String cacheKey = "hook:idem:" + idemKey + ":" + bodyHash;
        String seenHashKey = "hook:idem:hash:" + idemKey;

        // Cache hit → replay the first response verbatim; the side-effect does NOT run again.
        String cached = redis.opsForValue().get(cacheKey);
        if (cached != null) {
            return json(200, cached, true);
        }

        // Same key + different body → 409 (body-hash conflict gate).
        String priorHash = redis.opsForValue().get(seenHashKey);
        if (priorHash != null && !priorHash.equals(bodyHash)) {
            return json(409, "{\"statusCode\":409,\"message\":\"Idempotency-Key reused with a different body\"}", false);
        }

        // Demo retry: count attempts per key; fail the first N.
        String attemptKey = "hook:attempt:" + idemKey;
        Long attempt = redis.opsForValue().increment(attemptKey);
        redis.expire(attemptKey, Duration.ofSeconds(ttlSeconds));
        if (attempt != null && attempt <= failFirst) {
            return json(503, "{\"statusCode\":503,\"message\":\"simulated failure\"}", false);
        }

        // Cache miss → run the side-effect exactly once and capture its result.
        JsonNode bodyEcho = tryParse(rawBody);
        String eventId = (bodyEcho != null && bodyEcho.hasNonNull("id")) ? bodyEcho.get("id").asText() : idemKey;
        HookResult result = new HookResult(true, eventId, idemKey, Instant.now().toString(), bodyEcho);
        String payload = objectMapper.writeValueAsString(result);

        // SET ... NX EX: only the winner writes; a racing first-time request reads the
        // winner's value instead of double-running the side-effect.
        Boolean won = redis.opsForValue().setIfAbsent(cacheKey, payload, ttlSeconds, TimeUnit.SECONDS);
        redis.opsForValue().set(seenHashKey, bodyHash, ttlSeconds, TimeUnit.SECONDS);
        if (Boolean.FALSE.equals(won)) {
            String winner = redis.opsForValue().get(cacheKey);
            if (winner != null) return json(200, winner, true);
        }
        return json(200, payload, false);
    }

    private ResponseEntity<String> json(int status, String body, boolean replayed) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        if (replayed) headers.set("Idempotent-Replayed", "true");
        return ResponseEntity.status(status).headers(headers).body(body);
    }

    private JsonNode tryParse(String raw) {
        try {
            return objectMapper.readTree(raw);
        } catch (Exception e) {
            return null;
        }
    }

    private static String sha256(String s) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        byte[] digest = md.digest(s.getBytes(StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        for (byte b : digest) sb.append(String.format("%02x", b));
        return sb.toString();
    }
}
